crypto/objects/obj_xref.c

```diff
@@ -111,18 +148,49 @@ int OBJ_add_sigid(int signid, int dig_id, int pkey_id)
     ntr->hash_id = dig_id;
     ntr->pkey_id = pkey_id;
 
-    if (!sk_nid_triple_push(sig_app, ntr)) {
+    if (!CRYPTO_THREAD_write_lock(sig_lock)) {
+        ERR_raise(ERR_LIB_OBJ, ERR_R_LOCK_ACQUISITION_FAILED);
         OPENSSL_free(ntr);
         return 0;
     }
 
-    if (!sk_nid_triple_push(sigx_app, ntr))
-        return 0;
+    /* Check that the entry doesn't exist or exists as desired */
+    if (OBJ_find_sigid_algs(signid, &dnid, &pnid)) {
```

Comment:
Doesn't this cause a recursive lock? This call does read lock on sig_lock but you already have write lock.

Reply:
It does, I'll remove it.  There was no check originally.

Reply:
Put it back but with conditional locking.